### PR TITLE
ES6 Plugin: URL

### PR DESCRIPTION
**Pull request by iisa**
- move file to src
- light linting & conversion
- tests

#### To Test:
- pull branch
- update your modules, cmd: `npm install`
- run local server, cmd: `npm run serve`
- go to: http://127.0.0.1:8000/BookReaderDemo/demo-ia-olivertwist.html
- change page: URL updates page in path
- change view mode: URL updates view mode in path
- search inside book: URL updates search term in path

Internal ticket: https://webarchive.jira.com/browse/WEBDEV-3217

### CODE REVIEW

File: src/js/plugins/plugin.url.js [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+  const { urlMode, urlTrackIndex0, urlTrackedParams } = this.options;
+
+  if (!urlTrackIndex0
+      && (typeof(allParams.index) !== 'undefined')
+      && allParams.index === 0) {
+    delete allParams.index;
+    delete allParams.page;
+  }
+
+  const params = urlTrackedParams.reduce((validParams = {}, aParam) => {
```
Comment (cdrini) [inline src/js/plugins/plugin.url.js]: No need for default here.

```suggestion
  const params = urlTrackedParams.reduce((validParams, aParam) => {
```

File: src/js/plugins/plugin.url.js [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+    if (aParam in allParams) {
+      validParams[aParam] = allParams[aParam];
+    }
+    return validParams
+  }, {});
```
Comment (cdrini) [inline src/js/plugins/plugin.url.js]:
```suggestion
  const params = urlTrackedParams.reduce((validParams = {}, paramName) => {
    if (paramName in allParams) {
      validParams[paramName] = allParams[paramName];
    }
    return validParams;
  }, {});
```

File: src/js/plugins/plugin.url.js [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+  this.locationPollId = setInterval(updateHash, 500);
+};
+
+/**
+ * Update URL from the current parameters.
+ * Call this instead of manually using window.location.replace
+ */
+BookReader.prototype.urlUpdateFragment = function() {
+  const allParams = this.paramsFromCurrent();
+  const { urlMode, urlTrackIndex0, urlTrackedParams } = this.options;
```
Comment (cdrini) [inline src/js/plugins/plugin.url.js]: <3 I loooove this! So much easier to read, and makes the options this depends on _super_ clear.

File: src/js/plugins/plugin.url.js [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+/**
+ * Will read either the hash or URL and return the bookreader fragment
+ * @return {string}
+ */
+BookReader.prototype.urlReadFragment = function() {
+  if (this.options.urlMode === 'history') {
+    return window.location.pathname.substr(this.options.urlHistoryBasePath.length);
+  } else {
+    return window.location.hash.substr(1);
+  }
```
Comment (cdrini) [inline src/js/plugins/plugin.url.js]:
```suggestion
  const { urlMode, urlHistoryBasePath } = this.options;
  if (urlMode === 'history') {
    return window.location.pathname.substr(urlHistoryBasePath.length);
  } else {
    return window.location.hash.substr(1);
  }
```

File: src/js/plugins/plugin.url.js [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+  urlMode: 'hash',
+
+  // When using 'history' mode, this part of the URL is kept constant
+  // Example: /details/plato/
+  urlHistoryBasePath: '/',
+
+  // Only these params will be reflected onto the URL
+  urlTrackedParams: ['page', 'search', 'mode', 'region', 'highlight'],
+
+  // If true, don't update the URL when page == n0 (eg "/page/n0")
```
[the remark this answers is not in the document]
Comment (iisa) [reply src/js/plugins/plugin.url.js]: thanks so much!

File: src/js/plugins/plugin.url.js (hunk 4 above)
Comment (iisa) [reply src/js/plugins/plugin.url.js]: :)

**Pull request merged**